plugin/pkg/admission/defaultingressclass/admission.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package defaultingressclass
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	networkingv1beta1 "k8s.io/api/networking/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apiserver/pkg/admission"
+	genericadmissioninitializer "k8s.io/apiserver/pkg/admission/initializer"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/informers"
+	networkingv1beta1listers "k8s.io/client-go/listers/networking/v1beta1"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/apis/networking"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+const (
+	// PluginName is the name of this admission controller plugin
+	PluginName = "DefaultIngressClass"
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		plugin := newPlugin()
+		return plugin, nil
+	})
+}
+
+// claimDefaulterPlugin holds state for and implements the admission plugin.
+type claimDefaulterPlugin struct {
+	*admission.Handler
+	lister networkingv1beta1listers.IngressClassLister
+}
+
+var _ admission.Interface = &claimDefaulterPlugin{}
+var _ admission.MutationInterface = &claimDefaulterPlugin{}
+var _ = genericadmissioninitializer.WantsExternalKubeInformerFactory(&claimDefaulterPlugin{})
+
+// newPlugin creates a new admission plugin.
+func newPlugin() *claimDefaulterPlugin {
+	return &claimDefaulterPlugin{
+		Handler: admission.NewHandler(admission.Create),
+	}
+}
+
+// SetExternalKubeInformerFactory sets a lister and readyFunc for this
+// claimDefaulterPlugin using the provided SharedInformerFactory.
+func (a *claimDefaulterPlugin) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DefaultIngressClass) {
+		return
+	}
+	informer := f.Networking().V1beta1().IngressClasses()
+	a.lister = informer.Lister()
+	a.SetReadyFunc(informer.Informer().HasSynced)
+}
+
+// ValidateInitialization ensures lister is set.
+func (a *claimDefaulterPlugin) ValidateInitialization() error {
+	if a.lister == nil {
```

Comment:
return early if feature is off or this will block startup (start up an API server with `--runtime-config=extensions/v1beta1=false,networking.k8s.io/v1beta1=false --feature-gates=DefaultIngressClass=false` to verify this is functional)

Reply:
I've added another early return here if the feature gate is disabled and verified that API Server works with this combination of flags on a small e2e cluster. The `DefaultIngressClass` e2e tests expectedly fail on that cluster, but other basic actions seem to work well.